// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: c1/connector/v2/annotation_baton_id.proto

package v2

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ExternalResourceMatch with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ExternalResourceMatch) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExternalResourceMatch with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ExternalResourceMatchMultiError, or nil if none found.
func (m *ExternalResourceMatch) ValidateAll() error {
	return m.validate(true)
}

func (m *ExternalResourceMatch) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ResourceType

	// no validation rules for Key

	// no validation rules for Value

	if len(errors) > 0 {
		return ExternalResourceMatchMultiError(errors)
	}

	return nil
}

// ExternalResourceMatchMultiError is an error wrapping multiple validation
// errors returned by ExternalResourceMatch.ValidateAll() if the designated
// constraints aren't met.
type ExternalResourceMatchMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExternalResourceMatchMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExternalResourceMatchMultiError) AllErrors() []error { return m }

// ExternalResourceMatchValidationError is the validation error returned by
// ExternalResourceMatch.Validate if the designated constraints aren't met.
type ExternalResourceMatchValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExternalResourceMatchValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExternalResourceMatchValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExternalResourceMatchValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExternalResourceMatchValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExternalResourceMatchValidationError) ErrorName() string {
	return "ExternalResourceMatchValidationError"
}

// Error satisfies the builtin error interface
func (e ExternalResourceMatchValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExternalResourceMatch.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExternalResourceMatchValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExternalResourceMatchValidationError{}

// Validate checks the field values on ExternalResourceMatchAll with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ExternalResourceMatchAll) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExternalResourceMatchAll with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ExternalResourceMatchAllMultiError, or nil if none found.
func (m *ExternalResourceMatchAll) ValidateAll() error {
	return m.validate(true)
}

func (m *ExternalResourceMatchAll) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ResourceType

	if len(errors) > 0 {
		return ExternalResourceMatchAllMultiError(errors)
	}

	return nil
}

// ExternalResourceMatchAllMultiError is an error wrapping multiple validation
// errors returned by ExternalResourceMatchAll.ValidateAll() if the designated
// constraints aren't met.
type ExternalResourceMatchAllMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExternalResourceMatchAllMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExternalResourceMatchAllMultiError) AllErrors() []error { return m }

// ExternalResourceMatchAllValidationError is the validation error returned by
// ExternalResourceMatchAll.Validate if the designated constraints aren't met.
type ExternalResourceMatchAllValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExternalResourceMatchAllValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExternalResourceMatchAllValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExternalResourceMatchAllValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExternalResourceMatchAllValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExternalResourceMatchAllValidationError) ErrorName() string {
	return "ExternalResourceMatchAllValidationError"
}

// Error satisfies the builtin error interface
func (e ExternalResourceMatchAllValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExternalResourceMatchAll.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExternalResourceMatchAllValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExternalResourceMatchAllValidationError{}

// Validate checks the field values on BatonID with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *BatonID) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on BatonID with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in BatonIDMultiError, or nil if none found.
func (m *BatonID) ValidateAll() error {
	return m.validate(true)
}

func (m *BatonID) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return BatonIDMultiError(errors)
	}

	return nil
}

// BatonIDMultiError is an error wrapping multiple validation errors returned
// by BatonID.ValidateAll() if the designated constraints aren't met.
type BatonIDMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m BatonIDMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m BatonIDMultiError) AllErrors() []error { return m }

// BatonIDValidationError is the validation error returned by BatonID.Validate
// if the designated constraints aren't met.
type BatonIDValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e BatonIDValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e BatonIDValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e BatonIDValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e BatonIDValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e BatonIDValidationError) ErrorName() string { return "BatonIDValidationError" }

// Error satisfies the builtin error interface
func (e BatonIDValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sBatonID.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = BatonIDValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = BatonIDValidationError{}
